### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by fledith**
[DEMO LINK](https://fledith.github.io/react_todo-app-with-api/)

### CODE REVIEW

**Review by oleksandr-kotliarov — CHANGES_REQUESTED** [review]
image

---------------------------------------------------------------------------------------------------

image

---------------------------------------------------------------------------------------------------

- Hide everything except the NewTodoField if there are no todos yet;
- Don't hide everything if todos are just filtered out.
image

---------------------------------------------------------------------------------------------------

- if new title is the same as the old one just cancel editing;
- if after editing title is empty todo should be deleted (check example)

------------------------------------------------------------------------------------


When i try to edit again i see the same text that was before error
1. image
2.  image
3.  image

---------------------------------------------------------------------------------------------------

- And i dont know if its important but when i start to click fast enter after todo add, i receive a lot of todos, you can just disable form

File: src/App.tsx [hunk 1]
```diff
@@ -1,223 +1,313 @@
...
+
+    todos.forEach(todo => {
+      if (todo.completed) {
+        changedTodoId.push(todo.id);
+      }
+    });
+
+    setChangedTodosId(changedTodoId);
+
+    todos.forEach(todo => todo.completed && removeTodo(todo.id));
```
Comment (VETALEK) [inline src/App.tsx:185]: let's try `Promise.all` here
image

File: src/App.tsx [hunk 2]
```diff
@@ -1,223 +1,313 @@
...
+
+    updateTodoById(todoId, data)
+      .then(() => setShouldUpdate(true))
+      .catch(() => onError('Unable to update a todo'))
+      .finally(() => {
+        setIsLoading(false);
+      });
+  };
+
+  const toggleAllTodo = () => {
```
Comment (VETALEK) [inline src/App.tsx:153]: let's wrap your methods with `useCallback`

**Review by VETALEK — CHANGES_REQUESTED** [review]
GJ 👍 
commented

P.S.
when I click `ENTER` with empty name field it shows error only for a brief moment, should be 3 seconds

P.P.S.
i can add a todos with the space as a title

File: src/api/todos.ts [hunk 3]
```diff
@@ -5,4 +5,22 @@ export const getTodos = (userId: number) => {
   return client.get<Todo[]>(`/todos?userId=${userId}`);
 };
 
-// Add more methods here
+export const createTodo = (
+  title: string,
+  userId: number,
+  completed: boolean,
```
Comment (solaryasha) [inline src/api/todos.ts:11]: It's better to pass object, so you will never make mistake with wrong arguments order

File: src/components/TodoItem/TodoItem.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+interface Props {
+  todo: Todo;
+  selectedTodoId: number | null;
+  onDeleteTodo: (todoId: number) => void;
+  onUpdateTodo: (todoId: number, data: {}) => void;
+  isLoading: boolean;
+  changedTodosId: number[];
+}
+
+export const TodoItem = (props: Props) => {
```
Comment (solaryasha) [inline src/components/TodoItem/TodoItem.tsx:17]: use `React.memo`

File: src/components/TodoList/TodoList.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+interface Props {
+  todos: Todo[];
+  selectedTodoId: number | null;
+  onDeleteTodo: (todoId: number) => void;
+  onUpdateTodo: (todoId: number, data: {}) => void;
+  isLoading: boolean;
+  changedTodosId: number[];
+}
+
+export const TodoList = (props: Props) => {
```
Comment (solaryasha) [inline src/components/TodoList/TodoList.tsx:13]: use React.memo

File: src/App.tsx [hunk 6]
```diff
@@ -1,223 +1,356 @@
...
+  const [selectedTodoId, setSelectedTodoId] = useState<number | null>(null);
+  const [shouldUpdate, setShouldUpdate] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+  const [changedTodosId, setChangedTodosId] = useState<number[]>([]);
+  const [filteredBy, setFilteredBy] = useState<string | null>(null);
+
+  const onError = useCallback((errorTitle: string) => {
+    setErrorMessage(errorTitle);
+
+    setTimeout(() => setErrorMessage(''), 3000);
```
Comment (solaryasha) [inline src/App.tsx:30]: You should clear previous timer before starting new one

**Review by solaryasha — CHANGES_REQUESTED** [review]

File: src/api/todos.ts (hunk 3 above)
Comment (fledith) [reply src/api/todos.ts:11]: mb, but i dont wanna create new interface for object without id.

**Pull request closed without merging**